In [ ]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

In [12]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
df=pd.read_csv("/kaggle/input/web-page-phishing-dataset/web-page-phishing.csv")
#df.columns
#df['phishing'].value_counts()
X=df.drop('phishing',axis=1)  #Target Column:phishing (1=phishing,0=legitimate)
Y=df['phishing']
#X.shape,Y.shape
X_train,X_test,Y_train,Y_test=train_test_split(
    X,Y, test_size=0.2,random_state=42
)
model=DecisionTreeClassifier(random_state=42)
model.fit(X_train,Y_train)
Y_pred=model.predict(X_test)
#accuracy=accuracy_score(Y_test,Y_pred)
#print("Accuracy:",accuracy)
cm=confusion_matrix(Y_test,Y_pred)
cm
#print(classification_report(Y_test,Y_pred))

array([[11750,   948],
       [ 1393,  5925]])

## Model Used
- Decision Tree Classifier
## Reason
- Works well with tabular data
- Fast to train
- Baseline model for comparison
## Model Evaluation
- Accuracy: 88.3%
- Baseline model without hyperparameter tuning
- Model shows strong initial performance on phishing detection

Q. Why are false negatives more dangerous than false positives in phishing detection?
A. False negatives are more dangerous 

In [13]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
df=pd.read_csv("/kaggle/input/web-page-phishing-dataset/web-page-phishing.csv")
X=df.drop('phishing',axis=1)  #Target Column:phishing (1=phishing,0=legitimate)
Y=df['phishing']
X_train,X_test,Y_train,Y_test=train_test_split(
    X,Y, test_size=0.2,random_state=42
)
rf=RandomForestClassifier(
    n_estimators=200,
    random_state=42,
    n_jobs=-1,
    class_weight='balanced'
)
rf.fit(X_train,Y_train)
rf_pred=rf.predict(X_test)
rf_acc=accuracy_score(Y_test,rf_pred)
confusion_matrix(Y_test,rf_pred)
print("Random Forest Accuracy",rf_acc)
#print(classification_report(Y_test,rf_pred))

Random Forest Accuracy 0.8890387689848122


## Model Comparison Table
| Model           | Accuracy |False Negatives|
|-----------------|----------|---------------|
| Decision Tree   | 88.3%    |  1393         |
| Random Forest   | 88.9%    |  890          |

### The missing phishing sites were reduced by 36%.
## Model Comparison Insight
Two models were evaluated for phishing website detection:

- Decision Tree Classifier
- Random Forest Classifier with class-weight balancing

While the Decision Tree achieved comparable accuracy, 
the Random Forest with class-weight balancing significantly reduced 
false negatives (missed phishing websites), which is critical in 
security-sensitive applications. Therefore, Random Forest was 
selected as the preferred model.